File: script/level.title.js
define('level.title', ['sound', 'images'], function(sound, images) {

    function LevelTitle(src, duration, sound) {
        this.sound = sound;
        this.duration = duration;
        this.ttl = duration;
        var me = this;
        this.img = null;
        images.waitFor(src).done(function(img) {
            me.img = img;
        });
    }

    LevelTitle.prototype.draw = function(ctx) {
        ctx.fillStyle = '#111';
        ctx.fillRect(0, 0, ctx.canvas.width, ctx.canvas.height);
        if (!this.img) return;

        var hw = this.img.width / 2;
        var hh = this.img.height / 2;
        ctx.drawImage(
            this.img,
            ctx.canvas.width / 2 - hw,
            ctx.canvas.height / 2 - hh
        );
    };
    LevelTitle.prototype.init = function() {
        if (this.sound) {
            sound.play(this.sound);
        }
        this.ttl = this.duration;
    };
    LevelTitle.prototype.tick = function(delta, levelComplete) {
        this.ttl -= delta;
        if (this.ttl <= 0) {
            levelComplete();
        }
    };

    return {
        get: function(src, duration, sound) {
            return new LevelTitle(src, duration, sound);
        }
    }

});

File: script/level.title.dart
import 'dart:html';

import 'images.dart' as images;
import 'level.generic.dart';
import 'sound.dart' as sound;


class LevelTitle extends Level {

    images.Drawable image;
    int duration;
    String soundName;

    int ttl;

    LevelTitle(String src, int duration, {String soundName: null}) {
        this.image = images.get(src);
        this.duration = duration;
        this.soundName = soundName;
    }

    void reset() {
        if (this.soundName != null) {
            sound.play(this.soundName);
        }

        this.ttl = this.duration;
    }

    void draw(CanvasRenderingContext2D ctx, Function drawUI) {
        ctx.fillStyle = '#111';
        ctx.fillRect(0, 0, ctx.canvas.width, ctx.canvas.height);

        this.image.draw((img) {
            var hw = img.width / 2;
            var hh = img.height / 2;
            ctx.drawImage(
                img,
                ctx.canvas.width / 2 - hw,
                ctx.canvas.height / 2 - hh
            );
        });
    }

    void tick(int delta, Function nextLevel) {
        this.ttl -= delta;
        if (this.ttl <= 0) {
            nextLevel();
        }
    }

}
